diff --git a/alacritty/src/display/window.rs b/alacritty/src/display/window.rs
index 962f93a1db..f1f92b62e9 100644
--- a/alacritty/src/display/window.rs
+++ b/alacritty/src/display/window.rs
@@ -179,7 +179,4 @@ impl Window {
         window.set_transparent(config.window_opacity() < 1.);
 
-        #[cfg(target_os = "macos")]
-        use_srgb_color_space(&window);
-
         #[cfg(all(feature = "x11", not(any(target_os = "macos", windows))))]
         if !is_wayland {
